Issue with `aiida.manage.fixtures.FixtureManager.reset_db()` and testing WorkChains

In [this file](https://github.com/chrisjsewell/aiida-crystal17/blob/de068ac762d969725e8dd3a1d306dd1845f56a95/aiida_crystal17/tests/test_workflows/test_crystal_main.py) I have two tests that call `run_get_node` on a `WorkChain` (a replica of your `PwBaseWorkChain` thanks @shuber!). If I run either test individually, they work fine. But, if I run them in the same session, the 2nd one fails. Also, if I comment out `reset_db()` and run the test file, both tests pass.

```python

aiida_crystal17/tests/test_workflows/test_crystal_main.py:120: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../aiida_core_develop/aiida/engine/launch.py:62: in run_get_node
    return runner.run_get_node(process, *args, **inputs)
../aiida_core_develop/aiida/engine/runners.py:244: in run_get_node
    result, node = self._run(process, *args, **inputs)
../aiida_core_develop/aiida/engine/runners.py:210: in _run
    process = instantiate_process(self, process, *args, **inputs)
../aiida_core_develop/aiida/engine/utils.py:61: in instantiate_process
    process = process_class(runner=runner, inputs=inputs)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/state_machine.py:187: in __call__
    inst.transition_to(inst.create_initial_state())
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/state_machine.py:326: in transition_to
    self.transition_failed(initial_state_label, label, *sys.exc_info()[1:])
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/state_machine.py:339: in transition_failed
    six.reraise(type(exception), exception, trace)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/six.py:693: in reraise
    raise value
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/state_machine.py:310: in transition_to
    self._enter_next_state(new_state)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/state_machine.py:370: in _enter_next_state
    self._fire_state_event(StateEventHook.ENTERING_STATE, next_state)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/state_machine.py:288: in _fire_state_event
    callback(self, hook, state)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/processes.py:306: in <lambda>
    lambda _s, _h, state: self.on_entering(state))
../aiida_core_develop/aiida/engine/processes/process.py:327: in on_entering
    super(Process, self).on_entering(state)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/processes.py:599: in on_entering
    call_with_super_check(self.on_create)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/plumpy/base/utils.py:29: in call_with_super_check
    fn(*args, **kwargs)
../aiida_core_develop/aiida/engine/processes/process.py:323: in on_create
    self._pid = self._create_and_setup_db_record()
../aiida_core_develop/aiida/engine/processes/process.py:521: in _create_and_setup_db_record
    self._setup_db_record()
../aiida_core_develop/aiida/engine/processes/process.py:632: in _setup_db_record
    self._setup_inputs()
../aiida_core_develop/aiida/engine/processes/process.py:666: in _setup_inputs
    self.node.add_incoming(node, LinkType.INPUT_WORK, name)
../aiida_core_develop/aiida/orm/nodes/node.py:769: in add_incoming
    self._add_incoming_cache(source, link_type, link_label)
../aiida_core_develop/aiida/orm/nodes/node.py:827: in _add_incoming_cache
    if link_triple in self._incoming_cache:
../aiida_core_develop/aiida/orm/nodes/data/base.py:69: in __eq__
    return self.value == other
../aiida_core_develop/aiida/orm/nodes/data/base.py:57: in value
    return self.get_attribute('value', None)
../aiida_core_develop/aiida/orm/nodes/node.py:349: in get_attribute
    attribute = self.backend_entity.get_attribute(key)
../aiida_core_develop/aiida/orm/implementation/django/nodes.py:180: in get_attribute
    return self._dbmodel.attributes[key]
../aiida_core_develop/aiida/orm/implementation/django/utils.py:58: in __getattr__
    self._ensure_model_uptodate(fields=(item,))
../aiida_core_develop/aiida/orm/implementation/django/utils.py:141: in _ensure_model_uptodate
    self._model.refresh_from_db(fields=fields)
//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/django/db/models/base.py:697: in refresh_from_db
    db_instance = db_instance_qs.get()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <QuerySet []>, args = (), kwargs = {}, clone = <QuerySet []>, num = 0

    def get(self, *args, **kwargs):
        """
        Performs the query and returns a single object matching the given
        keyword arguments.
        """
        clone = self.filter(*args, **kwargs)
        if self.query.can_filter() and not self.query.distinct_fields:
            clone = clone.order_by()
        num = len(clone)
        if num == 1:
            return clone._result_cache[0]
        if not num:
            raise self.model.DoesNotExist(
                "%s matching query does not exist." %
>               self.model._meta.object_name
            )
E           aiida.backends.djsite.db.models.DoesNotExist: DbNode matching query does not exist.

//anaconda/envs/aiida-workshop/lib/python3.6/site-packages/django/db/models/query.py:380: DoesNotExist
```

--- a/aiida/engine/processes/calcjobs/calcjob.py
+++ b/aiida/engine/processes/calcjobs/calcjob.py
@@ -148,12 +148,12 @@ class CalcJob(Process):
             help='Set the quality of service to use in for the queue on the remote computer')
         spec.input('metadata.options.withmpi', valid_type=bool, default=False,
             help='Set the calculation to use mpi',)
-        spec.input('metadata.options.mpirun_extra_params', valid_type=(list, tuple), default=[],
+        spec.input('metadata.options.mpirun_extra_params', valid_type=(list, tuple), default=lambda: [],
             help='Set the extra params to pass to the mpirun (or equivalent) command after the one provided in '
                  'computer.mpirun_command. Example: mpirun -np 8 extra_params[0] extra_params[1] ... exec.x',)
         spec.input('metadata.options.import_sys_environment', valid_type=bool, default=True,
             help='If set to true, the submission script will load the system environment variables',)
-        spec.input('metadata.options.environment_variables', valid_type=dict, default={},
+        spec.input('metadata.options.environment_variables', valid_type=dict, default=lambda: {},
             help='Set a dictionary of custom environment variables for this calculation',)
         spec.input('metadata.options.priority', valid_type=str, required=False,
             help='Set the priority of the job to be queued')

--- a/aiida/engine/processes/ports.py
+++ b/aiida/engine/processes/ports.py
@@ -8,9 +8,9 @@
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
 """AiiDA specific implementation of plumpy Ports and PortNamespaces for the ProcessSpec."""
-
 import collections
 import re
+import warnings
 
 from plumpy import ports
 
@@ -95,6 +95,22 @@ class InputPort(WithSerialize, WithNonDb, ports.InputPort):
     Sub class of plumpy.InputPort which mixes in the WithSerialize and WithNonDb mixins to support automatic
     value serialization to database storable types and support non database storable input types as well.
     """
+
+    def __init__(self, *args, **kwargs):
+        """Override the constructor to check the type of the default if set and warn if not immutable."""
+        # pylint: disable=redefined-builtin,too-many-arguments
+        from aiida.orm import Node
+
+        if 'default' in kwargs:
+            default = kwargs['default']
+            # If the default is specified and it is a node instance, raise a warning. This is to try and prevent that
+            # people set node instances as defaults which can cause various problems.
+            if default is not ports.UNSPECIFIED and isinstance(default, Node):
+                message = 'default of input port `{}` is a `Node` instance, which can lead to unexpected side effects.'\
+                    ' It is advised to use a lambda instead, e.g.: `default=lambda: orm.Int(5)`.'.format(args[0])
+                warnings.warn(UserWarning(message))  # pylint: disable=no-member
+
+        super(InputPort, self).__init__(*args, **kwargs)
 
     def get_description(self):
         """

--- a/tests/engine/test_process.py
+++ b/tests/engine/test_process.py
@@ -317,7 +317,7 @@ class TestProcess(AiidaTestCase):
             @classmethod
             def define(cls, spec):
                 super().define(spec)
-                spec.input('add_outputs', valid_type=orm.Bool, default=orm.Bool(False))
+                spec.input('add_outputs', valid_type=orm.Bool, default=lambda: orm.Bool(False))
                 spec.output_namespace('integer.namespace', valid_type=orm.Int, dynamic=True)
                 spec.output('required_string', valid_type=orm.Str, required=True)
 

--- a/tests/engine/test_process_builder.py
+++ b/tests/engine/test_process_builder.py
@@ -31,7 +31,7 @@ class ExampleWorkChain(WorkChain):
         spec.input('name.spaced', valid_type=orm.Int, help='Namespaced port')
         spec.input('name_spaced', valid_type=orm.Str, help='Not actually a namespaced port')
         spec.input('boolean', valid_type=orm.Bool, help='A pointless boolean')
-        spec.input('default', valid_type=orm.Int, default=orm.Int(DEFAULT_INT).store())
+        spec.input('default', valid_type=orm.Int, default=lambda: orm.Int(DEFAULT_INT).store())
 
 
 class LazyProcessNamespace(Process):

--- a/tests/engine/test_process_function.py
+++ b/tests/engine/test_process_function.py
@@ -52,7 +52,7 @@ class TestProcessFunction(AiidaTestCase):
             return data_a
 
         @workfunction
-        def function_args_with_default(data_a=orm.Int(DEFAULT_INT)):
+        def function_args_with_default(data_a=lambda: orm.Int(DEFAULT_INT)):
             return data_a
 
         @calcfunction
@@ -72,12 +72,12 @@ class TestProcessFunction(AiidaTestCase):
             return result
 
         @workfunction
-        def function_args_and_default(data_a, data_b=orm.Int(DEFAULT_INT)):
+        def function_args_and_default(data_a, data_b=lambda: orm.Int(DEFAULT_INT)):
             return {'data_a': data_a, 'data_b': data_b}
 
         @workfunction
         def function_defaults(
-            data_a=orm.Int(DEFAULT_INT), metadata={
+            data_a=lambda: orm.Int(DEFAULT_INT), metadata={
                 'label': DEFAULT_LABEL,
                 'description': DEFAULT_DESCRIPTION
             }

--- a/tests/engine/test_work_chain.py
+++ b/tests/engine/test_work_chain.py
@@ -7,12 +7,12 @@
 # For further information on the license, see the LICENSE.txt file        #
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
-
 import inspect
 import unittest
 
 import plumpy
 import plumpy.test_utils
+import pytest
 from tornado import gen
 
 from aiida import orm
@@ -87,8 +87,8 @@ class Wf(WorkChain):
     @classmethod
     def define(cls, spec):
         super().define(spec)
-        spec.input('value', default=Str('A'))
-        spec.input('n', default=Int(3))
+        spec.input('value', default=lambda: Str('A'))
+        spec.input('n', default=lambda: Int(3))
         spec.outputs.dynamic = True
         spec.outline(
             cls.s1,
@@ -155,8 +155,8 @@ class PotentialFailureWorkChain(WorkChain):
     def define(cls, spec):
         super().define(spec)
         spec.input('success', valid_type=Bool)
-        spec.input('through_return', valid_type=Bool, default=Bool(False))
-        spec.input('through_exit_code', valid_type=Bool, default=Bool(False))
+        spec.input('through_return', valid_type=Bool, default=lambda: Bool(False))
+        spec.input('through_exit_code', valid_type=Bool, default=lambda: Bool(False))
         spec.exit_code(cls.EXIT_STATUS, 'EXIT_STATUS', cls.EXIT_MESSAGE)
         spec.outline(if_(cls.should_return_out_of_outline)(return_(cls.EXIT_STATUS)), cls.failure, cls.success)
         spec.output(cls.OUTPUT_LABEL, required=False)
@@ -943,7 +943,7 @@ class TestWorkChainAbortChildren(AiidaTestCase):
         @classmethod
         def define(cls, spec):
             super().define(spec)
-            spec.input('kill', default=Bool(False))
+            spec.input('kill', default=lambda: Bool(False))
             spec.outline(cls.begin, cls.check)
 
         def begin(self):
@@ -961,7 +961,7 @@ class TestWorkChainAbortChildren(AiidaTestCase):
         @classmethod
         def define(cls, spec):
             super().define(spec)
-            spec.input('kill', default=Bool(False))
+            spec.input('kill', default=lambda: Bool(False))
             spec.outline(cls.submit_child, cls.check)
 
         def submit_child(self):
@@ -1275,7 +1275,7 @@ class TestWorkChainExpose(AiidaTestCase):
                 }
             })
 
-    @unittest.skip('Reenable when issue #2515 is solved: references to deleted ORM instances')
+    @unittest.skip('Functionality of `WorkChain.exposed_outputs` is broken.')
     def test_nested_expose(self):
         res = launch.run(
             GrandParentExposeWorkChain,
@@ -1312,17 +1312,22 @@ class TestWorkChainExpose(AiidaTestCase):
                 }
             })
 
+    @pytest.mark.filterwarnings('ignore::UserWarning')
     def test_issue_1741_expose_inputs(self):
-        """Test that expose inputs works correctly when copying a stored default value"""
+        """Test that expose inputs works correctly when copying a stored default value.
 
-        stored_a = Int(5).store()
+        .. note:: a node instance is used for a port default, which is normally not advisable, but given that this
+            regression test relies on the default being a stored node, we cannot change it. Given that the default is
+            only used within this test, it should not pose any problems.
+
+        """
 
         class Parent(WorkChain):
 
             @classmethod
             def define(cls, spec):
                 super().define(spec)
-                spec.input('a', default=stored_a)
+                spec.input('a', default=Int(5).store())
                 spec.outline(cls.step1)
 
             def step1(self):
@@ -1401,7 +1406,7 @@ class TestDefaultUniqueness(AiidaTestCase):
         @classmethod
         def define(cls, spec):
             super().define(spec)
-            spec.input('a', valid_type=Bool, default=Bool(True))
+            spec.input('a', valid_type=Bool, default=lambda: Bool(True))
 
         def run(self):
             pass

--- a/tests/orm/test_querybuilder.py
+++ b/tests/orm/test_querybuilder.py
@@ -131,8 +131,8 @@ class TestQueryBuilder(AiidaTestCase):
             def define(cls, spec):
                 super().define(spec)
                 spec.input('success', valid_type=orm.Bool)
-                spec.input('through_return', valid_type=orm.Bool, default=orm.Bool(False))
-                spec.input('through_exit_code', valid_type=orm.Bool, default=orm.Bool(False))
+                spec.input('through_return', valid_type=orm.Bool, default=lambda: orm.Bool(False))
+                spec.input('through_exit_code', valid_type=orm.Bool, default=lambda: orm.Bool(False))
                 spec.exit_code(cls.EXIT_STATUS, 'EXIT_STATUS', cls.EXIT_MESSAGE)
                 spec.outline(if_(cls.should_return_out_of_outline)(return_(cls.EXIT_STATUS)), cls.failure, cls.success)
                 spec.output(cls.OUTPUT_LABEL, required=False)

